### PR TITLE
Account utilities for normalization and heuristic based correction of mnemonic phrases

### PR DESCRIPTION
**Pull request by nategraf**
### Description

Adds improved normalization for BIP-39 mnemonic phrases, as well as a new heuristic-based method for generating suggested corrections to a malformed mnemonic phrase.

Autocorrect suggestions are designed to correct words that may have been copied incorrectly, or mistyped when entered to recover an account. It suggests similar phrases, ordered by estimated probability that they are the source phrase of the observed phrase after being passed through a "noisy channel". Ordering is determined by edit total edit distance of the observed (incorrect) phrase to the suggestion.

### Other changes

* Major refactoring of `accounts.test.ts`
* Refactoring of `accounts.ts`

### Tested

Added unit tests which include a number of new phrases for normalization and correction testing.

### Related issues

- Related https://github.com/celo-org/celo-monorepo/issues/7060

### CODE REVIEW

**Comment by nategraf** [comment]
Looks like `formatNonAccentedCharacters` broke the `@celo/celocli` build. I'll got ahead and update usage of that function to use `normalizeMnemonic` instead.

File: packages/sdk/utils/src/account.ts [hunk 1]
```diff
@@ -55,70 +56,301 @@ export async function generateMnemonic(
...
+) {
+  if (language !== undefined) {
+    return bip39ToUse.validateMnemonic(mnemonic, getWordList(language))
+  }
+
   const languages = getAllLanguages()
-  for (const language of languages) {
-    if (bip39ToUse.validateMnemonic(mnemonic, getWordList(language))) {
+  for (const guessedLanguage of languages) {
+    if (bip39ToUse.validateMnemonic(mnemonic, getWordList(guessedLanguage))) {
```
Comment (jmrossy) [inline packages/sdk/utils/src/account.ts:70]: Maybe validation is super fast so it's irrelevant but I wonder if comparing the words to the language dictionaries would be quicker?

Comment (jmrossy) [reply packages/sdk/utils/src/account.ts:70]: Or actually I see you have a detectMnemonicLanguage function below. Why not use that here?

File: packages/sdk/utils/tsconfig.json [hunk 2]
```diff
@@ -2,7 +2,8 @@
   "extends": "@celo/typescript/tsconfig.library.json",
   "compilerOptions": {
     "rootDir": "src",
-    "outDir": "lib"
+    "outDir": "lib",
+    "downlevelIteration": true
```
Comment (jmrossy) [inline packages/sdk/utils/tsconfig.json:6]: Is this for handling accented chars better? I wish json allowed comments

**Review by jmrossy — COMMENTED** [review]
Overall lgtm, just a few questions below

File: packages/sdk/utils/tsconfig.json (hunk 2 above)
Comment (nategraf) [reply packages/sdk/utils/tsconfig.json:6]: It's to allow for iterating over generator functions

File: packages/sdk/utils/src/account.ts (hunk 1 above)
Comment (nategraf) [reply packages/sdk/utils/src/account.ts:70]: Its probably slightly faster to use `bip39.validateMnemonic` because that function will bail on the first invalid word instead of comparing the phrase to all languages. Mostly though, this way of doing it feel more direct and makes sense to me personally.

File: packages/sdk/utils/src/account.test.ts [hunk 3]
```diff
@@ -1,284 +1,499 @@
...
+          expected:
+            'déclarer effrayer estime carbone bebe danger déphaser junior buisson ériger morceau cintrer',
         },
-        derivation1: {
-          address: '0x773d283771e56abcef5925F9F5Aadb49FB91D647',
-          privateKey: '717dd0b70395964bc4175492f61599e7165cfbc6290fa76798fd77f8dfd9bbb9',
-          publicKey: '0361d8adcac067bb2927d625e642af5f1f53914b102d0740ad97d103ea079a6ce4',
+        {
+          mnemonic:
+            'declarer effrayer estime carbone bebe danger déphaser junior buisson ériger morceau cintrer',
```
Comment (barbaraliau) [inline packages/sdk/utils/src/account.test.ts:186]: I'm not sure I understand this test (probably because I don't know French or Spanish very well). The mnemonic is in French but it is expected to be in Spanish, so it adds the accents?

File: packages/sdk/utils/src/account.test.ts [hunk 4]
```diff
@@ -1,284 +1,499 @@
...
+
+    it('should normalize capitalized words to lowercase', async () => {
+      const mnemonic =
+        'female cousin RAPID exotic ribbon level equiP LeGal fuN RIVER hotel duTy TRIP youth rebel'
+      const expected =
+        'female cousin rapid exotic ribbon level equip legal fun river hotel duty trip youth rebel'
+      expect(normalizeMnemonic(mnemonic)).toEqual(expected)
+    })
+
+    it('should normalize extra and non-standard whitespace', async () => {
```
Comment (barbaraliau) [inline packages/sdk/utils/src/account.test.ts]: Are there any weird edge cases for whitespace with other languages like Korean or Japanese?

**Review by barbaraliau — APPROVED** [review]
LGTM. Couple questions about test cases, but no real concerns.

File: packages/sdk/utils/src/account.test.ts (hunk 3 above)
Comment (nategraf) [reply packages/sdk/utils/src/account.test.ts:186]: Added a comment to explain what this is testing

File: packages/sdk/utils/src/account.test.ts (hunk 4 above)
Comment (nategraf) [reply packages/sdk/utils/src/account.test.ts]: Indeed! Only Japanese, but I've added a test case specifically for it now.
